engine/app/components/citizens_advice_components/text_input.html.haml

```diff
@@ -1,2 +1,2 @@
-= render CitizensAdviceComponents::Input.new(base_input_args) do
+= render CitizensAdviceComponents::Input.new(**base_input_args) do
```

Comment:
@luke-hill This feels like something you might know? I _think_ I understand why I need the splats here, but trying to read up on the differences in keyword arguments in Ruby 3 made my head hurt. Any background here?

Reply:
So basically the issue is ambiguity. The biggest issue is that in ruby certain things are inferred or automatic (i.e. every method has a block argument, but often it is nil).

The problem is if you have a method with positional / named / hash arguments you need a way of differentiating them. Often people used to cheat and pass a hash of the named arguments. However this would work in **both instances**

i.e.
```
def foo(a: 1, b: 2)
  "a is #{a}, b is #{b}"
end
```

Here we have 2 keyword arguments so there's no real issue. Furthermore if you pass in a hash, it'll try find the keywords a and b, again no real issue.

However if you have a bit of a ambiguous one you can get issues. A simple (passing one), is the following

```
def bar(a = 1, **opts)
  "a is #{a}, opts is #{opts}"
end
```

Here we want a positional argument first, then a hash of all remaining options, with this method signature we have no issues until we pass in 2 things that look like a hash i.e.

```
bar("x" => 3, y: 5) # This gives different answers on ruby 2.6 and ruby 2.7
```

Because we're not sure whether this is one hash (And one argument), and should go as `a` or it's two hashes and 2 arguments and should overwrite a=1 and leave y:5 as the opts.

The full descriptive article (From the horses mouth), is this one. https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

However in short (TL;DR)

1) Nothing changes if you **only** have properly defined methods and you switch to using **kwargs on invocations.
2) If you expect a hash of options passed in. Be more specific now, don't be lazy
3) If you have a method that has positional, named and then a hash of options at the end, things may be different from ruby 2.6->2.7 (2.7 to 3 basically goes from throwing warnings to crashing).

A shorter explanation of the final point 3 is (From 2.6->2.7)

See https://github.com/site-prism/site_prism/pull/49 & https://www.ruby-lang.org/en/news/2019/12/25/ruby-2-7-0-released/

> When a method call passes a Hash at the last argument, and when it passes no keywords, and when the called method accepts keywords, a warning is emitted. To continue treating the hash as keywords, add a double splat operator to avoid the warning and ensure correct behavior in Ruby 3.